jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/execution/sam/TemplateUtils.kt

```diff
@@ -0,0 +1,92 @@
+// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+@file:JvmName("TemplateUtils")
+
+package software.aws.toolkits.jetbrains.services.lambda.execution.sam
+
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.vfs.LocalFileSystem
+import com.intellij.openapi.vfs.VirtualFile
+import com.intellij.psi.util.PsiUtil
+import org.jetbrains.yaml.psi.YAMLDocument
+import org.jetbrains.yaml.psi.YAMLFile
+import org.jetbrains.yaml.psi.YAMLKeyValue
+import org.jetbrains.yaml.psi.YAMLMapping
+import org.jetbrains.yaml.psi.YAMLPsiElement
+import software.aws.toolkits.resources.message
+import java.io.File
+
+private const val FUNCTION_TYPE = "AWS::Serverless::Function"
+
+data class SamFunction(
+    val logicalName: String,
+    val handler: String,
+    val runtime: String
+) {
+    override fun toString() = logicalName
+}
+
+fun findSamFunctionsFromTemplate(project: Project, file: File): List<SamFunction> {
+    val virtualFile = LocalFileSystem.getInstance().refreshAndFindFileByIoFile(file) ?: throw RuntimeException(
+        message("lambda.sam.template_not_found", file)
+    )
+    return findSamFunctionsFromTemplate(project, virtualFile)
+}
+
+fun findSamFunctionsFromTemplate(project: Project, file: VirtualFile): List<SamFunction> {
+    val yamlFile = PsiUtil.getPsiFile(project, file) as? YAMLFile ?: throw RuntimeException(message("lambda.sam.template_not_yaml", file.path))
+    return yamlFile.documents.flatMap {
+        it.findResourceByType(FUNCTION_TYPE)
+    }.mapNotNull {
+        val handler = it.getScalarProperty("Handler") ?: return@mapNotNull null
+        val runtime = it.getScalarProperty("Runtime") ?: return@mapNotNull null
+        SamFunction(logicalName = it.logicalName, handler = handler, runtime = runtime)
+    }
+}
+
+fun updateCodeUriForFunctions(file: File, newCodeUri: String) {
+    val regex = """CodeUri:.*""".toRegex()
+    val content = file.readText()
+    file.writeText(regex.replace(content, "CodeUri: $newCodeUri"))
+}
+
+fun functionFromElement(element: YAMLPsiElement): SamFunction? {
+    val keyValue = element as? YAMLKeyValue ?: return null
+    val value = keyValue.value as? YAMLMapping ?: return null
+    val type = value.getKeyValueByKey("Type")?.valueText ?: return null
+    if (type != FUNCTION_TYPE) {
```

Comment:
Should we support AWS::Lambda::Function too, since that can also be used inside a SAM template?